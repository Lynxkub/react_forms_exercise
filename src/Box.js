import React from 'react';


const Box = ({height , width , backgroundColor , deleteBox , id}) => {
        const newHeight = height + 'px';
        const newWidth = width + 'px'

        const handleClick = (e) => {
            // e.target.parentElement.isvisible = false;
            deleteBox(id)
        }
   return (
       <div>
       <div  style = { {height : newHeight , width : newWidth , backgroundColor : backgroundColor , display : 'inline-block'}}>
       </div>
        {height !== 0 && width !== 0 ? <button onClick = {handleClick}>X</button> : ""}
       </div>
   )
}

export default Box;

