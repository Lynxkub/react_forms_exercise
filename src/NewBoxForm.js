import React , {useState} from 'react';



const NewBoxForm = ({addBox}) => {

    const INITIAL_STATE = {
        width : "",
        height : "",
        backgroundColor : ""
    }

    const [formData , setFormData] = useState(INITIAL_STATE);

    const handleChange = (e) => {
        const {name , value } = e.target;

        setFormData(formData => ({
            ...formData,
            [name] : value
        }))
    } 

    const handleSubmit = (e) => {
        e.preventDefault();
        addBox({...formData})
        setFormData(INITIAL_STATE)
    }

    return (
        <div>
            <form onSubmit={handleSubmit}>
                <label htmlFor ='height'>Height</label>
                <input 
                data-testid='height'
                id='height'
                type = 'number'
                placeholder='height'
                name = 'height'
                value = {formData.height}
                onChange = {handleChange}
                />
                <label htmlFor = 'width'>Width</label>
                <input 
                data-testid='width'
                id = 'width'
                type = 'number'
                placeholder='width'
                name = 'width'
                value = {formData.width}
                onChange = {handleChange}
                />
                <label htmlFor ='backgroundColor'>Background Color</label>
                <input 
                data-testid='backgroundColor'
                id = 'backgroundColor'
                type = 'text'
                placeholder='background color'
                name = 'backgroundColor'
                value = {formData.backgroundColor}
                onChange = {handleChange}
                />
                <button>Add Box</button>
            </form>
        </div>
    )
}


export default NewBoxForm;