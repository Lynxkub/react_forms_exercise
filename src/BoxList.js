import React , {useState} from 'react'
import NewBoxForm from './NewBoxForm';
import Box from './Box'
import uuid from 'react-uuid';

const BoxList = () => {
    const INITIAL_STATE = [{
        height: 0,
        width : 0,
        backgroundColor: '',
        id : ""
    }]
    const [boxes , setBoxes] = useState(INITIAL_STATE);

    const addBox = (newBox) => {
        setBoxes(boxes => [...boxes , {...newBox , id : uuid() }])
        
    }


    const deleteBox = (id) => {
        setBoxes(boxes => boxes.filter(box => box.id !== id))
    }

    return (
        <div>
            <NewBoxForm addBox = {addBox}/>
            <h3>Created Boxes</h3>
            <div>
                {boxes.map(box => <Box deleteBox ={deleteBox} height = {box.height} width = {box.width} backgroundColor = {box.backgroundColor} key = {box.id} id = {box.id}/>)}
            </div>
        </div>
    )
}


export default BoxList;