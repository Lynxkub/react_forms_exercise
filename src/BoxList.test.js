import React from 'react';
import {render , fireEvent} from '@testing-library/react'
import Box from './Box';
import BoxList from './BoxList';
import NewBoxForm from './NewBoxForm';
import '@testing-library/jest-dom/extend-expect';


it('renders wihtout crashing' , function () {
    render(<BoxList />)
    render(<NewBoxForm />)
})


it('matches snapshot' , function () {
    const {asFragment} = render(<BoxList />)
    expect(asFragment()).toMatchSnapshot();
})

it('should create a box when form is filled out' , function () {
    const{queryByText, getByTestId} = render(<BoxList />);
    const height = getByTestId('height');
    const width = getByTestId('width');
    const color = getByTestId('backgroundColor');
    const addBox = queryByText('Add Box');
    // Before the box is created
    expect(queryByText('X')).not.toBeInTheDocument();
    fireEvent.change(height , {target : {value : 100}});
    fireEvent.change(width , {target : {value : 100}});
    fireEvent.change(color , {target : {value : 'blue'}});
    fireEvent.click(addBox);
    // Box is created along with a button with an X
    expect(queryByText('X')).toBeInTheDocument();
    // Deleting Box
    const deleteBox = queryByText('X');
    fireEvent.click(deleteBox);
    expect(queryByText('X')).not.toBeInTheDocument();
})