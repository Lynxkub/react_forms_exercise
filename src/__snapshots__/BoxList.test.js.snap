// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        <label
          for="height"
        >
          Height
        </label>
        <input
          data-testid="height"
          id="height"
          name="height"
          placeholder="height"
          type="number"
          value=""
        />
        <label
          for="width"
        >
          Width
        </label>
        <input
          data-testid="width"
          id="width"
          name="width"
          placeholder="width"
          type="number"
          value=""
        />
        <label
          for="backgroundColor"
        >
          Background Color
        </label>
        <input
          data-testid="backgroundColor"
          id="backgroundColor"
          name="backgroundColor"
          placeholder="background color"
          type="text"
          value=""
        />
        <button>
          Add Box
        </button>
      </form>
    </div>
    <h3>
      Created Boxes
    </h3>
    <div>
      <div>
        <div
          style="height: 0px; width: 0px; display: inline-block;"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
